CI: Conformance Gateway API: TestConformance/HTTPRouteServiceTypes

## CI failure

Sample failure can be found here https://github.com/cilium/cilium/actions/runs/9597945715/job/26468107591

Sysdump: [cilium-sysdump-out-false-standard.zip](https://github.com/user-attachments/files/15914649/cilium-sysdump-out-false-standard.zip)

```
--- FAIL: TestConformance/HTTPRouteServiceTypes (0.22s)

    conformance.go:64: 2024-06-20T13:06:53.078533807Z: Applying tests/httproute-service-types.yaml
    apply.go:273: 2024-06-20T13:06:53.081184685Z: Creating service-types HTTPRoute
    apply.go:273: 2024-06-20T13:06:53.089058603Z: Creating manual-endpointslices Service
    apply.go:273: 2024-06-20T13:06:53.10111978Z: Creating manual-endpointslices-ip4 EndpointSlice
    apply.go:273: 2024-06-20T13:06:53.120610512Z: Creating manual-endpointslices-ip6 EndpointSlice
    apply.go:273: 2024-06-20T13:06:53.132755546Z: Creating headless Service
    apply.go:273: 2024-06-20T13:06:53.142643629Z: Creating headless-manual-endpointslices Service
    apply.go:273: 2024-06-20T13:06:53.15435219Z: Creating headless-manual-endpointslices-ip4 EndpointSlice
    apply.go:273: 2024-06-20T13:06:53.16147056Z: Creating headless-manual-endpointslices-ip6 EndpointSlice
    helpers.go:611: 2024-06-20T13:06:53.176343149Z: Conditions matched expectations
    helpers.go:611: 2024-06-20T13:06:53.176399614Z: Route gateway-conformance-infra/service-types Parents matched expectations
    apply.go:281: 2024-06-20T13:07:23.250019968Z: Deleting headless-manual-endpointslices-ip6 EndpointSlice
    apply.go:281: 2024-06-20T13:07:23.253846345Z: Deleting headless-manual-endpointslices-ip4 EndpointSlice
    apply.go:281: 2024-06-20T13:07:23.257820386Z: Deleting headless-manual-endpointslices Service
    apply.go:281: 2024-06-20T13:07:23.265422036Z: Deleting headless Service
    apply.go:281: 2024-06-20T13:07:23.273158986Z: Deleting manual-endpointslices-ip6 EndpointSlice
    apply.go:281: 2024-06-20T13:07:23.27741143Z: Deleting manual-endpointslices-ip4 EndpointSlice
    apply.go:281: 2024-06-20T13:07:23.284735067Z: Deleting manual-endpointslices Service
    apply.go:281: 2024-06-20T13:07:23.300149767Z: Deleting service-types HTTPRoute

    --- FAIL: TestConformance/HTTPRouteServiceTypes/1_request_to_'/headless-manual-endpointslices'_should_go_to_infra-backend-v1 (30.00s)

    httproute-service-types.go:102: 2024-06-20T13:06:53.24931143Z: Making GET request to http://172.18.255.192/headless-manual-endpointslices
    http.go:240: 2024-06-20T13:06:53.266587743Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 500 (after 1.232µs)
    http.go:240: 2024-06-20T13:06:54.272235323Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 1.018284922s)
    http.go:240: 2024-06-20T13:06:55.275170805Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 2.023703386s)
    http.go:240: 2024-06-20T13:06:56.277613233Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 3.025949513s)
    http.go:240: 2024-06-20T13:06:57.280808664Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 4.029279772s)
    http.go:240: 2024-06-20T13:06:58.283124697Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 5.031563393s)
    http.go:240: 2024-06-20T13:06:59.285531474Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 6.033946912s)
    http.go:240: 2024-06-20T13:07:00.288664715Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 7.036933041s)
    http.go:240: 2024-06-20T13:07:01.291506941Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 8.040052133s)
    http.go:240: 2024-06-20T13:07:02.29434106Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 9.042985864s)
    http.go:240: 2024-06-20T13:07:03.29677718Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 10.045217812s)
    http.go:240: 2024-06-20T13:07:04.299756431Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 11.048210776s)
    http.go:240: 2024-06-20T13:07:05.302632303Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 12.051000797s)
    http.go:240: 2024-06-20T13:07:06.305468831Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 13.053577134s)
    http.go:240: 2024-06-20T13:07:07.308530866Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 14.056896621s)
    http.go:240: 2024-06-20T13:07:08.31120104Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 15.0598282s)
    http.go:240: 2024-06-20T13:07:09.313777006Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 16.062531179s)
    http.go:240: 2024-06-20T13:07:10.316018097Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 17.064626673s)
    http.go:240: 2024-06-20T13:07:11.318585757Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 18.066884242s)
    http.go:240: 2024-06-20T13:07:12.320741974Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 19.069356235s)
    http.go:240: 2024-06-20T13:07:13.323675178Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 20.072284474s)
    http.go:240: 2024-06-20T13:07:14.326048821Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 21.074601496s)
    http.go:240: 2024-06-20T13:07:15.328204291Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 22.076837889s)
    http.go:240: 2024-06-20T13:07:16.330781475Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 23.078986183s)
    http.go:240: 2024-06-20T13:07:17.333475581Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 24.081944246s)
    http.go:240: 2024-06-20T13:07:18.336216647Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 25.085008696s)
    http.go:240: 2024-06-20T13:07:19.3383551Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 26.086994651s)
    http.go:240: 2024-06-20T13:07:20.34084504Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 27.08939512s)
    http.go:240: 2024-06-20T13:07:21.343480115Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 28.091863153s)
    http.go:240: 2024-06-20T13:07:22.346182356Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.255.192 Path:/headless-manual-endpointslices RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be 200, got 503 (after 29.09485791s)
    http.go:221: 2024-06-20T13:07:23.249937424Z: timeout while waiting after 30 attempts, 0/3 successes

┌─────────────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                      PACKAGE                      │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼───────────────────────────────────────────────────┼───────┼──────┼──────┼───────│
│  FAIL   │ 129.07s │ github.com/cilium/cilium/operator/pkg/gateway-api │  --   │ 355  │  3   │  9    │
└─────────────────────────────────────────────────────────────────────────────────────────────────────┘
Error: Process completed with exit code 1.
```

--- a/.github/ariane-config.yaml
+++ b/.github/ariane-config.yaml
@@ -10,6 +10,7 @@ triggers:
     - conformance-ipsec-e2e.yaml
     - conformance-eks.yaml
     - conformance-externalworkloads.yaml
+    - conformance-gateway-api.yaml
     - conformance-ginkgo.yaml
     - conformance-ingress.yaml
     - conformance-multi-pool.yaml

--- a/.github/workflows/conformance-gateway-api.yaml
+++ b/.github/workflows/conformance-gateway-api.yaml
@@ -151,7 +151,6 @@ jobs:
 
           CILIUM_INSTALL_DEFAULTS="${{ steps.default_vars.outputs.cilium_install_defaults }} \
             --helm-set=kubeProxyReplacement=true \
-            --helm-set=extraArgs={--envoy-config-retry-interval=5s} \
             --helm-set=gatewayAPI.enabled=true \
             --helm-set=l2announcements.enabled=true"
 

--- a/pkg/ciliumenvoyconfig/cec_manager.go
+++ b/pkg/ciliumenvoyconfig/cec_manager.go
@@ -6,6 +6,7 @@ package ciliumenvoyconfig
 import (
 	"context"
 	"fmt"
+	"maps"
 	"slices"
 	"strconv"
 
@@ -28,7 +29,7 @@ type ciliumEnvoyConfigManager interface {
 	addCiliumEnvoyConfig(cecObjectMeta metav1.ObjectMeta, cecSpec *ciliumv2.CiliumEnvoyConfigSpec) error
 	updateCiliumEnvoyConfig(oldCECObjectMeta metav1.ObjectMeta, oldCECSpec *ciliumv2.CiliumEnvoyConfigSpec, newCECObjectMeta metav1.ObjectMeta, newCECSpec *ciliumv2.CiliumEnvoyConfigSpec) error
 	deleteCiliumEnvoyConfig(cecObjectMeta metav1.ObjectMeta, cecSpec *ciliumv2.CiliumEnvoyConfigSpec) error
-	syncCiliumEnvoyConfigService(name string, namespace string, cecSpec *ciliumv2.CiliumEnvoyConfigSpec) error
+	syncHeadlessService(name string, namespace string, serviceName loadbalancer.ServiceName, servicePorts []string) error
 }
 
 type cecManager struct {
@@ -153,43 +154,52 @@ func (r *cecManager) addK8sServiceRedirects(resourceName service.L7LBResourceNam
 			return err
 		}
 	}
-	return r.syncCiliumEnvoyConfigService(resourceName.Name, resourceName.Namespace, spec)
+	return r.syncAllHeadlessService(resourceName.Name, resourceName.Namespace, spec)
 }
 
-func (r *cecManager) syncCiliumEnvoyConfigService(name string, namespace string, spec *ciliumv2.CiliumEnvoyConfigSpec) error {
+func (r *cecManager) syncAllHeadlessService(name string, namespace string, spec *ciliumv2.CiliumEnvoyConfigSpec) error {
 	resourceName := service.L7LBResourceName{Name: name, Namespace: namespace}
 
 	// Register services for Envoy backend sync
 	for _, svc := range spec.BackendServices {
 		serviceName := getServiceName(resourceName, svc.Name, svc.Namespace, false)
-
-		if err := r.registerServiceSync(serviceName, resourceName, svc.Ports); err != nil {
+		if err := r.syncHeadlessService(name, namespace, serviceName, svc.Ports); err != nil {
 			return err
 		}
+	}
+	return nil
 
-		kSvc, err := r.getK8sService(svc.Name, svc.Namespace)
-		if err != nil {
+}
+
+func (r *cecManager) syncHeadlessService(name string, namespace string, serviceName loadbalancer.ServiceName, servicePorts []string) error {
+	resourceName := service.L7LBResourceName{Name: name, Namespace: namespace}
+
+	if err := r.registerServiceSync(serviceName, resourceName, servicePorts); err != nil {
+		return err
+	}
+
+	kSvc, err := r.getK8sService(serviceName.Name, serviceName.Namespace)
+	if err != nil {
+		return err
+	}
+
+	// Skip non-headless services as they are already supported in Cilium datapath LB service.
+	if kSvc == nil || kSvc.Spec.ClusterIP != slim_corev1.ClusterIPNone {
+		return nil
+	}
+
+	ep, err := r.getEndpoint(serviceName.Name, serviceName.Namespace)
+	if err != nil {
+		return err
+	}
+	if ep == nil {
+		return nil
+	}
+
+	// Explicitly call backendSyncer.Sync for headless service
+	for _, s := range convertToLBService(kSvc, ep) {
+		if err = r.backendSyncer.Sync(s); err != nil {
 			return err
-		}
-
-		// Skip non-headless services as they are already supported in Cilium datapath LB service.
-		if kSvc == nil || kSvc.Spec.ClusterIP != slim_corev1.ClusterIPNone {
-			continue
-		}
-
-		ep, err := r.getEndpoint(svc.Name, svc.Namespace)
-		if err != nil {
-			return err
-		}
-		if ep == nil {
-			continue
-		}
-
-		// Explicitly call backendSyncer.Sync for headless service
-		for _, s := range convertToLBService(kSvc, ep) {
-			if err = r.backendSyncer.Sync(s); err != nil {
-				return err
-			}
 		}
 	}
 	return nil
@@ -246,15 +256,22 @@ func (r *cecManager) getEndpoint(serviceName string, serviceNamespace string) (*
 		return nil, err
 	}
 
+	var res *k8s.Endpoints
 	iter := store.IterKeys()
 	for iter.Next() {
 		e, _, _ := store.GetByKey(iter.Key())
 		if e.EndpointSliceID.ServiceID.Name == serviceName &&
 			e.EndpointSliceID.ServiceID.Namespace == serviceNamespace {
-			return e, nil
+			if res == nil {
+				res = e.DeepCopy()
+			} else {
+				if len(e.Backends) > 0 {
+					maps.Insert(res.Backends, maps.All(e.Backends))
+				}
+			}
 		}
 	}
-	return nil, nil
+	return res, nil
 }
 
 func convertToLBService(svc *slim_corev1.Service, ep *k8s.Endpoints) []*loadbalancer.SVC {

--- a/pkg/ciliumenvoyconfig/cec_reconciler.go
+++ b/pkg/ciliumenvoyconfig/cec_reconciler.go
@@ -13,11 +13,13 @@ import (
 	"github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/cilium/cilium/pkg/k8s"
 	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
 	"github.com/cilium/cilium/pkg/k8s/synced"
+	"github.com/cilium/cilium/pkg/loadbalancer"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/node"
@@ -298,19 +300,32 @@ func (r *ciliumEnvoyConfigReconciler) configSelectsLocalNode(cfg *config) (bool,
 	return true, nil
 }
 
-func (r *ciliumEnvoyConfigReconciler) syncHeadlessService(_ context.Context) error {
+func (r *ciliumEnvoyConfigReconciler) syncEndpoints(_ context.Context, event resource.Event[*k8s.Endpoints]) error {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
 
 	var reconcileErr error
+	defer event.Done(reconcileErr)
 
-	for key, cfg := range r.configs {
-		if err := r.manager.syncCiliumEnvoyConfigService(cfg.meta.Name, cfg.meta.Namespace, cfg.spec); err != nil {
-			r.logger.WithField("key", key).WithError(err).Info("Failed to sync headless service, Hive will retry")
-			reconcileErr = errors.Join(reconcileErr, fmt.Errorf("failed to reconcile existing config (%s): %w", key, err))
-			continue
-		}
+	if event.Object == nil {
+		return reconcileErr
 	}
 
+	for key, cfg := range r.configs {
+		for _, svc := range cfg.spec.BackendServices {
+			if svc.Name != event.Object.EndpointSliceID.ServiceID.Name ||
+				svc.Namespace != event.Object.EndpointSliceID.ServiceID.Namespace {
+				continue
+			}
+
+			serviceName := loadbalancer.ServiceName{Name: svc.Name, Namespace: svc.Namespace}
+			if err := r.manager.syncHeadlessService(cfg.meta.Name, cfg.meta.Namespace, serviceName, svc.Ports); err != nil {
+				r.logger.WithField("key", key).
+					WithField(logfields.ServiceKey, event.Key).
+					WithError(err).Error("failed to sync headless service")
+				reconcileErr = errors.Join(reconcileErr, fmt.Errorf("failed to sync headless service (%s): %w", key, err))
+			}
+		}
+	}
 	return reconcileErr
 }

--- a/pkg/ciliumenvoyconfig/cec_reconciler_test.go
+++ b/pkg/ciliumenvoyconfig/cec_reconciler_test.go
@@ -20,6 +20,7 @@ import (
 	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/loadbalancer"
 	"github.com/cilium/cilium/pkg/node"
 	"github.com/cilium/cilium/pkg/node/types"
 )
@@ -789,6 +790,6 @@ func (r *fakeCECManager) updateCiliumEnvoyConfig(oldCECObjectMeta metav1.ObjectM
 	return nil
 }
 
-func (r *fakeCECManager) syncCiliumEnvoyConfigService(name string, namespace string, cecSpec *ciliumv2.CiliumEnvoyConfigSpec) error {
+func (r *fakeCECManager) syncHeadlessService(_ string, _ string, _ loadbalancer.ServiceName, _ []string) error {
 	return nil
 }

--- a/pkg/ciliumenvoyconfig/cell.go
+++ b/pkg/ciliumenvoyconfig/cell.go
@@ -15,6 +15,7 @@ import (
 	"github.com/cilium/cilium/pkg/envoy"
 	"github.com/cilium/cilium/pkg/k8s"
 	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/client"
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
 	"github.com/cilium/cilium/pkg/k8s/synced"
@@ -26,6 +27,12 @@ import (
 	"github.com/cilium/cilium/pkg/time"
 )
 
+// allService is a resource.Resource[*slim_corev1.Service]
+type allService resource.Resource[*slim_corev1.Service]
+
+// allEndpoint is a resource.Resource[*slim_corev1.Service]
+type allEndpoint resource.Resource[*k8s.Endpoints]
+
 // Cell provides support for the CRD CiliumEnvoyConfig that backs Ingress, Gateway API
 // and L7 loadbalancing.
 var Cell = cell.Module(
@@ -36,6 +43,14 @@ var Cell = cell.Module(
 	cell.ProvidePrivate(newCECManager),
 	cell.ProvidePrivate(newCECResourceParser),
 	cell.ProvidePrivate(newEnvoyServiceBackendSyncer),
+	cell.ProvidePrivate(
+		func(lc cell.Lifecycle, cfg k8s.Config, cs client.Clientset) (allService, error) {
+			return k8s.ServiceResource(lc, cfg, cs)
+		},
+		func(lc cell.Lifecycle, cfg k8s.Config, cs client.Clientset) (allEndpoint, error) {
+			return k8s.EndpointsResource(lc, cfg, cs)
+		},
+	),
 	cell.Config(cecConfig{}),
 )
 
@@ -66,6 +81,8 @@ type reconcilerParams struct {
 	CECResources   resource.Resource[*ciliumv2.CiliumEnvoyConfig]
 	CCECResources  resource.Resource[*ciliumv2.CiliumClusterwideEnvoyConfig]
 	LocalNodeStore *node.LocalNodeStore
+
+	EndpointResources allEndpoint
 }
 
 func registerCECK8sReconciler(params reconcilerParams) {
@@ -114,11 +131,13 @@ func registerCECK8sReconciler(params reconcilerParams) {
 	// Note: LocalNodeStore (in comparison to `resource.Resource`) doesn't provide a retry mechanism
 	params.JobGroup.Add(job.Observer("local-node-events", reconciler.handleLocalNodeEvent, params.LocalNodeStore))
 
+	// Observing service events for headless services
+	params.JobGroup.Add(job.Observer("headless-endpoint-events", reconciler.syncEndpoints, params.EndpointResources))
+
 	// TimerJob periodically reconciles all existing configs.
 	// This covers the cases were the reconciliation fails after changing the labels of a node.
 	if params.Config.EnvoyConfigRetryInterval > 0 {
 		params.JobGroup.Add(job.Timer("reconcile-existing-configs", reconciler.reconcileExistingConfigs, params.Config.EnvoyConfigRetryInterval))
-		params.JobGroup.Add(job.Timer("sync-headless-service", reconciler.syncHeadlessService, params.Config.EnvoyConfigRetryInterval))
 	}
 }
 
@@ -136,8 +155,8 @@ type managerParams struct {
 	BackendSyncer  *envoyServiceBackendSyncer
 	ResourceParser *cecResourceParser
 
-	Services  resource.Resource[*slim_corev1.Service]
-	Endpoints resource.Resource[*k8s.Endpoints]
+	Services  allService
+	Endpoints allEndpoint
 }
 
 func newCECManager(params managerParams) ciliumEnvoyConfigManager {

